### DISCUSSION TITLE
Clipping Path operations

### DISCUSSION THREAD

**Discussion by dorphalsig** [python-pillow/Pillow#6297] [Q&A] [answered]
Hi,  Imagemagick provides something called "clipping operations" with basically  operate on  a mask defined by a set of bezier curves. These curves are stored by photoshop inside the exif/jfif data of tiffs and jpgs.

Since I need to do this for a project, I'd also like to contribute the code, which brings me to the following questions:
- What would be needed to have the code merged? Do you guys only accept code that closes current issues, or do you also accept new functionality (like this)?
- What is your policy regarding dependencies? I need to render the mask as an image, which I'm thinking of doing by using some library to render SVGs as PNGs and then use these as the mask...

**Comment by radarhere** [comment, marked as answer]
We accept new features that don't just close current issues, yes.

There isn't a written stance on adding more dependencies that I can find. Obviously fewer is better, and ideally dependencies have a license such that we can bundle them into our binary wheels and distribute them.

If you're thinking about reading SVG files, you may want to also consider that we have an open issue requesting SVG support - #3509. Although just adding a dependency to convert SVGs to PNGs may not be an easy solution - once upon a time, I had a tentative idea to read PDFs on macOS by converting them to PNG, and that ended up being closed - https://github.com/python-pillow/Pillow/pull/1969#issuecomment-238076315

**Reply by dorphalsig** [reply]
So, I'm still trying to figure out if its easier to create the bezier curves directly in PIL by calculating as many points as it makes sense and then passing these points to pillow to create a polygon in a blank image that would act as the mask (the general idea is explained [here](https://note.nkmk.me/en/python-pillow-composite/). You most likely already know everything about that, but I'm posting it here also to kinda document the whole idea properly.  The points for the polygon would be calculated using something like [this](https://stackoverflow.com/questions/246525/how-can-i-draw-a-bezier-curve-using-pythons-pil), the problem is I'm not getting the right points :/

The other option is to give up on the idea of doing that and going for an SVG library that would give me directly an image I can use as the mask. This [SF post](https://stackoverflow.com/questions/69660200/how-to-render-svg-image-to-png-file-in-python) shows a line-up of 3 libraries that can do the trick: svglib, pycairo and inkscape (the latter not being a library but an external tool). In the post svglib and pycairo show similar results with a complex SVG, where inkscape rendered it exactly as expected (vs what a firefox and chrome render)- I just feel its kinda overkill to use inkscape for this because it would be a very primitive integration via temp files, and also the api that we could expose to interact with the svgs would be limited (because I'd have to build it basically from scratch, which let's face, is not going to happen). This would leave us with svgcairo which needs cairo installed or svglib, which depends on ReportLab ("commercial open source" python library that deals with PDFs and vector graphics (which seems to also fill a gap in pillow). I looked the open source license and its quite simple, allows for redistribution, modification and embedding as long as the copyright notice is maintained in the sources.

After all this, somehow I would be slightly inclined to try the SVG way, but I kinda need input here on what you guys think about the options

**Reply by radarhere** [reply]
From a maintenance perspective, if not perhaps from an initial implementation perspective, calculating bezier curve points seems far simpler to me than adding two dependencies and an external tool.

**Reply by dorphalsig** [reply]
hi, half a century later I revived and managed to get the whole clipping path thing working. As usual I was going the hard way around the thing.  The whole clipping path thing is around 100 lines long.  How should I commit it? Should I fork it and then do a PR and ref this comment?

**Reply by radarhere** [reply]
Sure, sounds good.
